### ISSUE TITLE
Standardize `Delegate`s

### ISSUE THREAD

**Issue by Desdaemon** [fzyzcjy/flutter_rust_bridge#1104]
**Is your feature request related to a problem? Please describe.**
The current design doesn't allow `List<String?>`, since in Rust this is `Vec<Option<String>>` and interactions between Option and Delegate are known to be problematic. Option requires that the underlying type provide an `api_fill_to_wire`, but no Delegates currently do this since they usually delegate to a List instead (String to StringList, Uuid to Uuids etc.)

**Describe the solution you'd like**
Since most delegates are currently simple type transformations, we can refactor `Delegate` as follows:

```rust
pub enum IrTypeDelegate {
    Simple {
        api: Box<IrType>,
        delegate: Box<IrType>,
        code: &'static dyn Delegate
    },
    // other methods of delegations
}

pub struct DelegateMode {
    pub to_api: bool,
    pub dart: bool,
}

pub trait Delegate {
    fn generate(&self, api: &IrType, delegate: &IrType, mode: DelegateMode) -> Option<String>;
}
```

**Describe alternatives you've considered**
Continue using the split Delegate/DelegateList design, but this is not ideal and it's hard to reason about.

**Additional context**
None.

**Comment by fzyzcjy** [comment]
Yes that looks like a problem. I am not very sure about what your full solution will be, but looks good with the high-level description and looking forward to your PR!
